### ISSUE TITLE
chefignore in at cookbook root not honored by `knife cookbook upload`

### ISSUE THREAD

**Issue by pburkholder** [chef/chef#2994]
This seems to be a change from 11.X.  Reported by @mcalb. To reproduce [see comment below, original comment deleted by @pburkholder since it was pretty unhelpful]

**Comment by pburkholder** [comment]
My apologies to the triage team for my lame report on a Friday evening.  Here's the README.md from https://github.com/pburkholder/chefignore-demo, which demonstrates this issue:

This cookbook demonstrates that chef-client 12.0.3 does not honor the contents
of `chefignore`, which is a security bug since content the creator considers
secret is uploaded to the chef-server in contravention to what one expects:


The `secrets` file in this cookbook has one secret:

    MySecretPassword

and the `chefignore` file has one entry:

    secrets

In chef-client 11.X, the `knife cookbook upload` command would not upload the
secrets file per the contents of chefignore, however, observe what happens

```
> knife --version
Chef: 12.0.3
> knife cookbook list | grep chefignore
> pwd
/Users/pburkholder/Hack/chefignore-demo
> knife cookbook upload -o ~/Hack chefignore-demo
Uploading chefignore-demo [0.1.0]
Uploaded 1 cookbook.
> cd ~/tmp
> knife cookbook download chefignore-demo
Downloading chefignore-demo cookbook version 0.1.0
Downloading resources
Downloading providers
Downloading recipes
Downloading definitions
Downloading libraries
Downloading attributes
Downloading files
Downloading templates
Downloading root_files
Cookbook downloaded to /Users/pburkholder/tmp/chefignore-demo-0.1.0
> more chefignore-demo-0.1.0/secrets
MySecretPassword
> more chefignore-demo-0.1.0/chefignore
secrets
>
```

Oops, so we've uploaded the secrets, and downloaded them, despite the
instructions of the `chefignore` file.

**Comment by pburkholder** [comment]
In case it's not clear above, here's the cookbook structure:

```
> tree .
.
├── README.md
├── chefignore
├── metadata.rb
├── recipes
│   └── default.rb
└── secrets

1 directory, 5 files
> more chefignore
secrets
```

Note that if you do, `cp chefignore ..` so the chefignore file is at the top of the cookbook_path, then the contents of chefignore are honored (a symlink also works).

**Comment by pburkholder** [comment]
https://github.com/chef/chef/blob/master/lib/chef/cookbook_loader.rb#L150 only seems look at repo_path, not in the cookbooks themselves. I hope to come back to this later.  P.

**Comment by err0** [comment]
👍  we just ran into this.  we run one repo per cookbook and expected the functionality to be the same as .gitignore.

**Comment by acoulton** [comment]
I've just come across this too - it also comes up when using chef in legacy solo mode. This results in no chefignore being applied when using the `SoloRunner` in chefspec and is partially responsible for [very slow specs if there are build-time dependencies or other additional files present in the cookbook path](https://github.com/chefspec/chefspec/issues/870).

The offending line is now at https://github.com/chef/chef/blob/master/lib/chef/cookbook_loader.rb#L168:

```ruby
def preload_cookbook(cookbook_path)
      repo_path = File.dirname(cookbook_path)
      @chefignores[repo_path] ||= Cookbook::Chefignore.new(repo_path)
      loader = Cookbook::CookbookVersionLoader.new(cookbook_path, @chefignores[repo_path])
      add_cookbook_loader(loader)
end
```

Given a cookbook_path of `/tmp/d20170811-18859-mdjbcf/cookbooks/ingenerator-source` this attempts to read a chefignore at `/tmp/d20170811-18859-mdjbcf/cookbooks/chefignore`. This doesn't exist, while the ignore at `/tmp/d20170811-18859-mdjbcf/cookbooks/ingenerator-source/cookbooks` is then skipped.

**Comment by briceo** [comment]
I have also hit this issue.  It seems like https://github.com/chef/chef/blob/master/lib/chef/cookbook_loader.rb#L174 is the line that needs changed:
`chefignore(File.dirname(cookbook_path))` to `chefignore(cookbook_path)`

I'm not sure what else that might affect, but in my situation, I was uploading a single cookbook repo to a supermarket server, and the `chefignore` file was being ignored (as it was looking in the parent directory rather than the cookbook directory).

**Comment by ronabop** [comment]
Similar boat here, one repo per cookbook pattern, Jenkins checking out a repo, trying to knife it up, chefignore inside the cookbook is ignored. (a ~/.chef/chefignore that worked could also fix my issue, but inside-a-cookbook-repo seems to be a very frequent use-case)

**Comment by vsingh-msys** [comment]
@ronabop We have provided the fixes in #8925 soon it will available.

**Comment by vsingh-msys** [comment]
Fixed in #8925
